### PR TITLE
  PR for review

### PR DESCRIPTION
**Pull request by vmwhoami**
The Weather App project comes at the end of the Asynchronous Javascript and APIs section of the Javascript course.
It tests the knowledge of the student about asynchronous communication with promises or async/await and when to use them.
Users can get weather information using API.

### CODE REVIEW

File: src/scripts/async.js [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+  try {
+    const response = await fetch(url);
+    const data = await response.json();
+    populateDom(data);
+    const { lat } = data.coord;
+    const { lon } = data.coord;
+    await getPredictions(lat, lon, 'minutely,hourly');
+  } catch (error) {
+    const err = document.querySelector('.error__msg');
+    err.textContent = "Couldn't find your location";
```
Comment (bellom) [inline src/scripts/async.js]: Let's separate DOM related tasks from logic tasks.

You can create another method to handle DOM tasks.

image

https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59622133/

File: README.md [hunk 2]
```diff
@@ -0,0 +1,57 @@
+# Weather-App
+
+Microverse project - Weather App
+
+## Project features
+
+- API Use example
+- The Weather App project comes at the end of the Asynchronous Javascript and APIs section of the Javascript course. It tests the knowledge of the student about asynchronous communication with promises or async/await and when to use them.
+
+## Live Demo Website
```
Comment (bellom) [inline README.md:10]: Hey, Seems you forgot to add the live demo link.
Kindly add it 👍 .

**Review by bellom — CHANGES_REQUESTED** [review]
## Hi @vmwhoami  :clap: 

### Awesome job so far!! :+1: 

#### You have done wonderfully well, you passed `almost` all the checklist for this project :+1: just a few issues left to work on............ :slightly_smiling_face: 

# Status: Changes Required! ♻️ 

**Kindly check out my comments below;**

### ♻️ Kindly re-submit for review after making all those changes :slightly_smiling_face: Thumps -up await you.....

**Best,**

**Oluwaseun Bello**
**_Slack Handle - @sean bellom_**

File: package.json [hunk 3]
```diff
@@ -14,6 +14,7 @@
   "devDependencies": {
     "babel-eslint": "^10.1.0",
     "css-loader": "^5.0.1",
+    "dotenv-webpack": "^6.0.0",
```
Comment (bellom) [inline package.json:18]: Inside this file.
we need to work on `script`
`npm start` seems not working.
Kindly double-check.

File: src/scripts/queries.js [hunk 4]
```diff
@@ -0,0 +1,7 @@
+
+const apikey = process.env.KEY;
+const forecast = (lat, lon, part = {}) => `https://api.openweathermap.org/data/2.5/onecall?lat=${lat}&lon=${lon}&exclude=${part}&appid=${apikey}`;
+const weatherNow = (location) => `http://api.openweathermap.org/data/2.5/weather?q=${location}&APPID=${apikey}`;
```
Comment (bellom) [inline src/scripts/queries.js:3]: Good job using `.env`
I think you don't need to use `.env`, this is just a weather app for you to display your knowledge on API calls.

**Review by bellom — CHANGES_REQUESTED** [review]
## Hi @vmwhoami  :clap: 

### Awesome job so far!! :+1: 

#### You have done wonderfully well, you passed `almost` all the checklist for this project :+1: just a few issues left to work on............ :slightly_smiling_face: 

# Status: Changes Required! ♻️ 

**Kindly check out my comments below;**

### ♻️ Kindly re-submit for review after making all those changes :slightly_smiling_face: Thumps -up await you.....

**Best,**

**Oluwaseun Bello**
**_Slack Handle - @sean bellom_**

File: src/scripts/async.js [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+export default async function fetchData(url) {
+  try {
+    const response = await fetch(url);
+    const data = await response.json();
+    populateDom(data);
+    const { lat } = data.coord;
+    const { lon } = data.coord;
+    await getPredictions(lat, lon, 'minutely,hourly');
+  } catch (error) {
+    showError(error);
```
Comment (mosaaleb) [inline src/scripts/async.js]: - [ ] This error message will never show up if the city is not found. The reason behind this is that the catch block will only catch errors due to a failure within fetch itself (network error, server not responding, ...). Any results come back from the server even if it fails to find a resource i.e 404 city not found won't be rejected. 
You can simulate a network error to check this in the dev console if you blocked the request domain:

Screenshot from 2020-12-20 18-09-44

To get this issue solved we need to check if the `response` is `ok` and not 404.
My suggestion is to make this function just return data  wether the response succeeded or failed

```js
  try {
    const response = await fetch(url);

    if (!response.ok) {
      // rejected promise => catch block
      // you will still get json data from weather API even if the city is not found
      const dataError = await response.json();
      return Promise.reject(new Error(dataError.message));
    }

    // resolved promise => then block
    const data = await response.json();
    return data;

    // Can be moved to the function call in then block, check my next comment
    // populateDom(data);
    // const { lat } = data.coord;
    // const { lon } = data.coord;
    // await getPredictions(lat, lon, 'minutely,hourly');
  } catch (error) {
    // rejected promise => catch block
    return Promise.reject(new Error('network failure'));
  }
```

But you are free to handle this issue how you see fit.

File: src/scripts/events.js [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+
+  window.onload = () => {
+    input.focus();
+  };
+
+  btn.addEventListener('click', () => {
+    const errmsg = inputValidtion(input);
+    if (!errmsg) {
+      const searchloc = weatherNow(input.value);
+      fetchData(searchloc);
```
Comment (mosaaleb) [inline src/scripts/events.js]: - [ ] Then we can use `then` and `catch` blocks here to continue our data flow:

```js
fetchData(searchloc)
  .then((res) => {
    /* resolved promise */
    // handle successful response
    // populateData function
    // other logic
  })
  .catch((err) => {
    /* rejected promise */
    // showError(err)
  });
```

Resources : 
[fetch resolves even if 404?](https://stackoverflow.com/questions/39297345/fetch-resolves-even-if-404)
[Fetch Async Wait](https://dmitripavlutin.com/javascript-fetch-async-await/) 
[Promise reject vs throw](https://stackoverflow.com/questions/33445415/javascript-promises-reject-vs-throw)

**Review by mosaaleb — CHANGES_REQUESTED** [review]
#### Changes required

Well done @vmwhoami :clap:. You have handled all the requirements from the previous review. Yet there is a bug that prevents showing the error message when the city is not found. Kindly take a look at my comments below.



- [ ] **[OPTIONAL]** it has a ‘loading’ component that displays from the time the form is submitted until the information comes back from the AP.




Please handle this requirement and submit a new code review.

##### Muhammad Ebeid

**Review by whiz25 — APPROVED** [review]
## Approved :clap:
Nice work, @vmwhoami. You have made all the requested changes. Your app correctly displays and error when a city is not found. Well done.
You may merge this PR and continue to the next project. Good luck.


### Martin Afani

**Pull request merged**